### PR TITLE
 Add support for opening  lemmy links

### PR DESCRIPTION
**Pull request by ggichure**
## Pull Request Description

Add support for opening Lemmy links in app  using [uni_links](https://pub.dev/packages/uni_links)

 - [x] Users e.g https://lemmy.world/u/lwadmin
 - [x] Posts  e.g https://lemmy.world/post/6740849
 - [x]  Comments  e.g https://lemmy.world/comments/1234567
 - [x] Depends on #818 
 - [x] I'll need help with the scripting part to auto-gen instances. #834 

Android support for now.

## Issue Being Fixed

Issue Number: #60 

## Screenshots / Recordings

|             |               |              |
|:--------------------------:|:--------------------------:|:--------------------------:|
| Image 1 | Image 2 | Image 3 |

## Checklist

- [x] Did you update CHANGELOG.md?
- [x] Did you use localized strings where applicable?
- [ ] Did you add `semanticLabel`s where applicable for accessibility?

### CODE REVIEW

**Comment by micahmo** [comment]
This is awesome, thanks for working on this!!

I know it's still a draft, so I didn't officially review, but I wanted to jot down some thoughts before I forgot. Note that I'm not asking for you to do any of these things. Just notes for reference. 😊

 - As you mentioned, it only supports Android for now (maybe @hjiangsu can implement for iOS as described [here](https://github.com/avioli/uni_links/blob/master/uni_links/README.md#for-ios)).
 - As you mentioned, the list of supported domains is hard-coded. We can help to auto-generate the list in the [instance workflow](https://github.com/thunder-app/thunder/blob/develop/.github/workflows/instances.yml).
 - Any reason for picking [uni_links](https://pub.dev/packages/uni_links) over [app_links](https://pub.dev/packages/app_links)? (I haven't researched either one so I'm asking purely out of curiosity.)
 - We should have graceful handling of any non-post link (I'm assuming we're only supporting posts to start).
 - Does the user need to manually enable link handling for each supported domain in the app's Android settings? (I believe that's required unless an app publisher can verify that it owns a domain.) If so, we should document that process.

Again, super awesome, and please don't take any comments as criticism or asking for extra work. 😊

**Comment by micahmo** [comment]
> > Does the user need to manually enable link handling for each supported domain in the app's Android settings? (I believe that's required unless an app publisher can verify that it owns a domain.) If so, we should document that process.
> 
> Given that we are using [deep links](https://developer.android.com/training/app-links/deep-linking) we don't have to verify we own the domains. Deep links are general links that point to specific content within an app.
> 
> [App links](https://developer.android.com/training/app-links) on the other hand are designed to verify and ensure that links are opened within the designated app rather than through a web browser, if the app is installed.

I understand that we don't need to (and can't) verify all the domains. However, I believe we still need to instruct users to enable link handling for all of those domains in the Android settings. For example, if I grab and run your branch, Thunder does not automatically handle those domains. I have to enable them, like this.

https://github.com/thunder-app/thunder/assets/7417301/18a21e2d-4640-4ead-b62f-f24f63a2c787

This is what I mean when I say we should document the process to enable link handling (unless I am misunderstanding something!).

> I thoroughly enjoy working on Thunder, and I welcome constructive criticism.

Me too. 😊 Thanks for all the help!

**Comment by ggichure** [comment]
> > > Does the user need to manually enable link handling for each supported domain in the app's Android settings? (I believe that's required unless an app publisher can verify that it owns a domain.) If so, we should document that process.
> > 
> > 
> > Given that we are using [deep links](https://developer.android.com/training/app-links/deep-linking) we don't have to verify we own the domains. Deep links are general links that point to specific content within an app.
> > [App links](https://developer.android.com/training/app-links) on the other hand are designed to verify and ensure that links are opened within the designated app rather than through a web browser, if the app is installed.
> 
> I understand that we don't need to (and can't) verify all the domains. However, I believe we still need to instruct users to enable link handling for all of those domains in the Android settings. For example, if I grab and run your branch, Thunder does not automatically handle those domains. I have to enable them, like this.
> qemu-system-x86_64_ow54t3xYYh.mp4
> 
> This is what I mean when I say we should document the process to enable link handling (unless I am misunderstanding something!).
> 
> > I thoroughly enjoy working on Thunder, and I welcome constructive criticism.
> 
> Me too. 😊 Thanks for all the help!


Yeah, we need to add a guide.

**Comment by micahmo** [comment]
Hey again! I opened PR #834 to help with auto-generating the supported domain list in the Android manifest file via the GitHub workflows. It assumes that the file already contains...

``` xml
...
<intent-filter>
  <action android:name="android.intent.action.VIEW" />
  <category android:name="android.intent.category.DEFAULT" />
  <category android:name="android.intent.category.BROWSABLE" />
  <data android:scheme="https"/>
   <!-- TODO(lemmy_server_domain_finder): Setup a script to update available lemmy domains  -->
  <!-- Obtained from github.com/dessalines/jerboa/blob/main/app/src/main/AndroidManifest.xml-->
  <!--#AUTO_GEN_INSTANCE_LIST_DO_NOT_TOUCH#-->
  ...
  <data android:host="WHATEVER" />
  ...
  <!--#INSTANCE_LIST_END#-->
</intent-filter>
...
```

We can either...
 - Copy my changes and incorporate them into your PR
 - Do my PR separately from your changes

I have no preference. :)

**Comment by ggichure** [comment]
> Hey again! I opened PR #834 to help with auto-generating the supported domain list in the Android manifest file via the GitHub workflows. It assumes that the file already contains...
> 
> ```
> ...
> <intent-filter>
>   <action android:name="android.intent.action.VIEW" />
>   <category android:name="android.intent.category.DEFAULT" />
>   <category android:name="android.intent.category.BROWSABLE" />
>   <data android:scheme="https"/>
>    
>   
>   
>   ...
>   <data android:host="WHATEVER" />
>   ...
>   
> </intent-filter>
> ...
> ```
> 
> We can either...
> 
>     * Copy my changes and incorporate them into your PR
> 
>     * Do my PR separately from your changes
> 
> 
> I have no preference. :)

Your PR can go in separately. 
Thanks for this.

**Comment by ggichure** [comment]
@micahmo  @hjiangsu  

I have a `commentId` , `navigateToPost` requires at least a `postId` or  `PostViewMedia` , any insights on how to navigate user to PostPage just using `commentId`?

**Comment by micahmo** [comment]
> I have a `commentId` , `navigateToPost` requires at least a `postId` or `PostViewMedia` , any insights on how to navigate user to PostPage just using `commentId`?

I had intended to do a followup to #818 by refactoring a `navigateToComment` method and using it for in-app navigation as well (you can see my `TODO` [here](https://github.com/thunder-app/thunder/blob/ae474735370a07390c3dd1327204997cef20d216/lib/shared/common_markdown_body.dart#L119)). Let me see if I can knock that out for you!

P.S. I was thinking you could use the instance page from #824 if they navigate to the root domain and maybe also if they navigate to an unsupported link.

**Comment by hjiangsu** [comment]
This is pretty awesome, thanks again for initiating work on this! I don't have too much experience with integrating deep links but I'll see what I can do on the iOS side of things to make it work as well.

> This is what I mean when I say we should document the process to enable link handling (unless I am misunderstanding something!).

I wonder if we can make it easier by having a button which opens up the app settings page for the user to be able to select the applicable domains? I'm not sure if this is possible but it would be a nice QoL feature!

**Comment by ggichure** [comment]
> I wonder if we can make it easier by having a button which opens up the app settings page for the user to be able to select the applicable domains? I'm not sure if this is possible but it would be a nice QoL feature!

Using `permission_handler` you can open app settings.

**Comment by micahmo** [comment]
> I don't have too much experience with integrating deep links but I'll see what I can do on the iOS side of things

Hopefully it should just involve creating/editing an entitlements file, and the code should be the same for both platforms.



> I wonder if we can make it easier by having a button which opens up the app settings page for the user to be able to select the applicable domains?

That's a great idea, and I'm pretty sure it's possible on Android!

**Comment by ggichure** [comment]
Open settings done. 
open_app_settings

**Comment by ggichure** [comment]
@micahmo  
`getLemmyCommentId` `ResolveObject` doesn't  "always" work  . [instance.dart#L114](https://github.com/micahmo/thunder/blob/f6362522c299357b6351c372300ccc7370fb46dd/lib/utils/instance.dart#L114)

Default instance being used lemmy.ml
On lemmy.world it works.
 https://lemmy.world/comment/3078559

```dart
 final ResolveObjectResponse resolveObjectResponse = await lemmy.run(ResolveObject(q: text));
``` 

> LemmyApiException: couldnt_find_object

**Comment by micahmo** [comment]
> Open settings done.

That's awesome! Is there any way to get it to the specific settings page for links? Maybe not with permission_manager but maybe with [app_settings](https://pub.dev/packages/app_settings)? Also do you want to add an icon to that setting so that it aligns with the other ones?

> `getLemmyCommentId` `ResolveObject` doesn't "always" work .

It won't be able to resolve an object that is not federated to the current instance. I'm not sure what we can do about that. Maybe just show a message? We could still open it in the browser in the app I suppose.

**Comment by ggichure** [comment]
[app_settings](https://pub.dev/packages/app_settings) seems like it will do the job.


> Maybe just show a message? We could still open it in the browser in the app I suppose.

Sure.
> Also do you want to add an icon to that setting so that it aligns with the other ones?

Sure

**Comment by micahmo** [comment]
> Using Universal Links means that we would need to verify the domains we add to the entitlements/info.plist files.

Ah man, that's too bad. I was hoping it would work like Android where the domains are simply disabled by default. But it makes sense from a security perspective why Apple would be a little more restrictive about this. Oh well!

**Review by micahmo — COMMENTED** [review]
Hey @ggichure, since you marked this ready for review, I went ahead and reviewed. Hope that's ok. 😊 Looks great overall! I left a couple comments on the code. I also took this for a test run.

The UI for the new setting still looks a bit funny, with the leading icon misaligned and the trailing very large. It also doesn't have the same stadium border and inkwell as the rest of the settings. Not sure if it's possible to reuse one of the existing setting widgets with a custom action. Also it still just opens to the main settings page (if you did not intend to change that in this PR, that's totally fine!!).

| image |
| - |

The main issue is that things don't seem to be working very well for me! I'm sure I'm doing something wrong, but here's what I'm seeing. This is with a post on the same instance that I'm logged in with, so shouldn't be a federation issue. (The same thing happens whether Thunder is already open or not.)

https://github.com/thunder-app/thunder/assets/7417301/46714b1b-b7fd-430a-931c-d450ee016c82

> P.S. Thanks as always for the additional localization.

File: lib/utils/navigate_post.dart
Comment (ggichure) [reply lib/utils/navigate_post.dart:63]: Since the post page  has them , I added them, I'm not so clear on how the post page works.

File: lib/utils/navigate_post.dart
Comment (ggichure) [reply lib/utils/navigate_post.dart:38]: The post page can work with either `PostviewMedia` or `postId`.

**Comment by ggichure** [comment]
> The UI for the new setting still looks a bit funny, with the leading icon misaligned and the trailing very large. It also doesn't have the same stadium border and inkwell as the rest of the settings. Not sure if it's possible to reuse one of the existing setting widgets with a custom action.

Reused the `ToggleOption` to create a `SettingsListTile`.
Screenshot 2023-10-19 at 11 29 28

**Comment by ggichure** [comment]
> The main issue is that things don't seem to be working very well for me! I'm sure I'm doing something wrong, but here's what I'm seeing. This is with a post on the same instance that I'm logged in with, so shouldn't be a federation issue. (The same thing happens whether Thunder is already open or not.)

I'm unable to reproduce this. 
Also I'm considering increasing the snackbar duration on link exception. .

**Comment by hjiangsu** [comment]
Just did a quick code review and everything seems to be fine with me, nothing too much to point out aside from what was already discussed previously (thanks @micahmo for doing a code review as well!)

I unfortunately can't test this myself since I don't have a physical Android device, so I'll leave it up to @micahmo to approve of this if thats okay

**Review by micahmo — APPROVED** [review]
If @ggichure can't reproduce the issue I was seeing, then I approve! Might as well get it into the nightly and get some feedback. 😊

**Comment by hjiangsu** [comment]
Sounds good! Just to confirm, @ggichure is this PR ready to go?

**Comment by ggichure** [comment]
> Sounds good! Just to confirm, @ggichure is this PR ready to go?

Yes it Is.

**Pull request merged**
